### PR TITLE
test: assert events emitted from contract 

### PR DESCRIPTION
**Pull request by chungquantin**
Link to issue: https://github.com/r0gue-io/pop-node/issues/181
Refactor fungible `integration-tests` contract to emit events by default and catch those events in the integration test cases.

### CODE REVIEW

**Review by peterwht — APPROVED** [review]

**Comment by chungquantin** [comment]
@Daanvdplas Could I receive review for this PR?

File: pop-api/integration-tests/src/fungibles/mod.rs [hunk 1]
```diff
@@ -364,8 +382,14 @@ fn instantiate_and_create_fungible_works() {
 			Err(Module { index: 52, error: 5 })
 		);
 		// Successfully create an asset when instantiating the contract.
-		assert_ok!(instantiate_and_create_fungible(contract, ASSET_ID, 1));
+		let instantiator =
+			instantiate_and_create_fungible(contract, ASSET_ID, 1).expect("Should work");
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/fungibles/mod.rs:386]: Could we remove the `expect` and make sure we `assert_ok` it and separately get contract address pls

File: pop-api/integration-tests/src/fungibles/mod.rs [hunk 2]
```diff
@@ -364,8 +382,14 @@ fn instantiate_and_create_fungible_works() {
...
 		assert!(Assets::asset_exists(ASSET_ID));
+		// Successfully emit an event from instantiating a new fungible.
```
Comment (Daanvdplas) [inline pop-api/integration-tests/src/fungibles/mod.rs:388]: Can we change all the event checks to the following: 
```suggestion
		// Successfully emit event.
```

One reason is simplification and the other reason is the use of the word "asset". The latter is something we don't want to anymore, always refer to "token".

File: pallets/api/src/fungibles/mod.rs [hunk 3]
```diff
@@ -275,7 +275,7 @@ pub mod pallet {
 			Ok(().into())
 		}
 
-		/// Decreases the allowance of a `spender` by `value` amount of tokens.
+		/// Decreases the allowance of a spender and asset.
```
Comment (Daanvdplas) [inline pallets/api/src/fungibles/mod.rs:278]: Can we remove this pls

**Review by Daanvdplas — COMMENTED** [review]
Looking good! Left small nitpicks. 

This PR is going to daan/api but I'm not sure whether we want this. I say we really focus on getting `frank/extension` merged and then rebase those small PRs.

File: pallets/api/src/fungibles/mod.rs (hunk 3 above)
Comment (evilrobot-01) [reply pallets/api/src/fungibles/mod.rs:278]: +1

File: pop-api/integration-tests/Cargo.toml [hunk 4]
```diff
@@ -17,23 +17,30 @@ frame-system = { version = "29.0.0", default-features = false }
...
 pallet-contracts = { version = "28.0.0", default-features = false }
+pop-api = { path = "../../pop-api", default-features = false, features = [
+    "fungibles",
+] }
 pop-primitives = { path = "../../primitives", default-features = false }
 pop-runtime-devnet = { path = "../../runtime/devnet", default-features = false }
 sp-io = { version = "31.0.0", default-features = false }
 sp-runtime = { version = "32.0.0", default-features = false }
 
+# TODO
```
Comment (evilrobot-01) [inline pop-api/integration-tests/Cargo.toml:28]: What is the todo?

File: pop-api/integration-tests/contracts/fungibles/lib.rs [hunk 5]
```diff
@@ -61,7 +64,13 @@ mod fungibles {
 
 		#[ink(message)]
 		pub fn transfer(&mut self, id: AssetId, to: AccountId, value: Balance) -> Result<()> {
-			api::transfer(id, to, value)
+			let result = api::transfer(id, to, value);
+			self.env().emit_event(Transfer {
```
Comment (evilrobot-01) [inline pop-api/integration-tests/contracts/fungibles/lib.rs:68]: Why emit the event if the transfer failed? Should be transfer(...)?; and then Ok(()) surely?

Comment (evilrobot-01) [reply pop-api/integration-tests/contracts/fungibles/lib.rs:68]: Same with all the others.

File: pop-api/integration-tests/contracts/fungibles/lib.rs [hunk 6]
```diff
@@ -7,7 +7,10 @@
 ///
 use ink::prelude::vec::Vec;
 use pop_api::{
-	assets::fungibles::{self as api},
+	assets::fungibles::{
+		self as api,
+		events::{Approval, ClearMetadata, Create, SetMetadata, StartDestroy, Transfer},
```
Comment (evilrobot-01) [inline pop-api/integration-tests/contracts/fungibles/lib.rs:12]: Generally accepted convention for event naming is past tense - i.e. Created, MetadataSet etc.

Assume this is based on the convention set by ERC20/PSP22

File: pop-api/integration-tests/src/fungibles/mod.rs [hunk 7]
```diff
@@ -345,6 +359,10 @@ fn create_works() {
 		assert_eq!(create(addr.clone(), ASSET_ID, BOB, 0), Err(Module { index: 52, error: 7 }),);
 		// Create asset successfully.
 		assert_ok!(create(addr.clone(), ASSET_ID, BOB, 1));
+		// Successfully emit an event from creating a new asset class.
```
Comment (evilrobot-01) [inline pop-api/integration-tests/src/fungibles/mod.rs:362]: Probably needs aligning with the token terminology PR.

File: pop-api/integration-tests/src/fungibles/mod.rs [hunk 8]
```diff
@@ -383,6 +407,9 @@ fn start_destroy_works() {
 		assert_eq!(start_destroy(addr.clone(), asset), Err(Module { index: 52, error: 2 }),);
 		let asset = create_asset(addr.clone(), ASSET_ID, 1);
 		assert_ok!(start_destroy(addr.clone(), asset));
+		// Successfully emit an event from destroying an asset.
```
Comment (evilrobot-01) [inline pop-api/integration-tests/src/fungibles/mod.rs:410]: Terminology. Please check the rest of the file/PR too.

File: pop-api/integration-tests/src/fungibles/utils.rs [hunk 9]
```diff
@@ -329,11 +329,30 @@ pub(super) fn instantiate_and_create_fungible(
...
+	let result = decoded::<Result<(), Error>>(result.clone().result.clone())
+		.unwrap_or_else(|_| panic!("Contract reverted: {:?}", result));
+	match result {
+		Ok(_) => Ok(account_id),
+		Err(error) => Err(error),
+	}
+}
+
+/// Get the latest event from pallet contracts.
+pub(super) fn latest_contract_event() -> Vec<u8> {
```
Comment (evilrobot-01) [inline pop-api/integration-tests/src/fungibles/utils.rs:346]:
```suggestion
pub(super) fn last_contract_event() -> Vec<u8> {
```

File: pop-api/integration-tests/src/fungibles/utils.rs [hunk 10]
```diff
@@ -329,11 +329,30 @@ pub(super) fn instantiate_and_create_fungible(
...
+	let account_id = result.clone().account_id;
+	let result = decoded::<Result<(), Error>>(result.clone().result.clone())
+		.unwrap_or_else(|_| panic!("Contract reverted: {:?}", result));
+	match result {
+		Ok(_) => Ok(account_id),
+		Err(error) => Err(error),
+	}
+}
+
+/// Get the latest event from pallet contracts.
```
Comment (evilrobot-01) [inline pop-api/integration-tests/src/fungibles/utils.rs:345]:
```suggestion
/// Get the last event from pallet contracts.
```

**Review by evilrobot-01 — CHANGES_REQUESTED** [review]
A few things need tweaking I'm afraid: return type of call not checked before emitting events, token terminology throughout (may want to get that PR merged first) and then it needs conflicts resolving.

File: pop-api/integration-tests/contracts/fungibles/lib.rs (hunk 6 above)
Comment (Daanvdplas) [reply pop-api/integration-tests/contracts/fungibles/lib.rs:12]: Correct

File: pop-api/src/primitives.rs [hunk 11]
```diff
@@ -1,5 +1,11 @@
...
+use ink::scale::Decode;
 pub use pop_primitives::*;
 
 pub(crate) type AccountId = <DefaultEnvironment as Environment>::AccountId;
 pub(crate) type Balance = <DefaultEnvironment as Environment>::Balance;
+
+/// Decode slice of bytes to environment associated type AccountId.
+pub fn account_id_from_slice(s: &[u8; 32]) -> AccountId {
+	AccountId::decode(&mut &s[..]).expect("Should be decoded to AccountId")
+}
```
Comment (Daanvdplas) [inline pop-api/src/primitives.rs:11]: in a `#[cfg(test)]` feature as it is only used for the integration tests.

Comment (Daanvdplas) [reply pop-api/src/primitives.rs:11]: Same for decode import

File: pop-api/integration-tests/contracts/create_token_in_constructor/lib.rs [hunk 12]
```diff
@@ -14,22 +14,23 @@ mod create_token_in_constructor {
...
-		id: TokenId,
+		token: TokenId,
 	}
 
 	impl Fungible {
 		#[ink(constructor, payable)]
-		pub fn new(id: TokenId, min_balance: Balance) -> Result<Self> {
-			let contract = Self { id };
+		pub fn new(token: TokenId, min_balance: Balance) -> Result<Self> {
+			let contract = Self { token };
```
Comment (Daanvdplas) [inline pop-api/integration-tests/contracts/create_token_in_constructor/lib.rs:23]: This should stay `id` as the token is not created yet. Same for in the contract struct

File: pop-api/integration-tests/contracts/fungibles/lib.rs [hunk 13]
```diff
@@ -131,16 +150,20 @@ mod fungibles {
 		#[ink(message)]
 		pub fn create(
 			&mut self,
-			id: TokenId,
+			token: TokenId,
```
Comment (Daanvdplas) [inline pop-api/integration-tests/contracts/fungibles/lib.rs:153]: should stay `id`

File: pop-api/src/primitives.rs (hunk 11 above)
Comment (Daanvdplas) [reply pop-api/src/primitives.rs:11]: We could also just put this function in integration tests?

File: pop-api/src/v0/fungibles.rs [hunk 14]
```diff
@@ -101,10 +101,10 @@ pub mod events {
 
 	/// Event emitted when a token is created.
 	#[ink::event]
-	pub struct Create {
+	pub struct Created {
 		/// The token identifier.
 		#[ink(topic)]
-		pub id: TokenId,
+		pub token: TokenId,
```
Comment (Daanvdplas) [inline pop-api/src/v0/fungibles.rs:107]: Should stay `id`

**Review by Daanvdplas — CHANGES_REQUESTED** [review]

**Review by Daanvdplas — APPROVED** [review]

**Review by evilrobot-01 — APPROVED** [review]
Looks good, pleased with the event naming.

**Pull request merged**